### PR TITLE
Fix/improve kubelet lifecycle handlers tests

### PR DESCRIPTION
**Pull request by PI-Victor**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
Improves testing coverage for lifecycle admission pod handlers.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: pkg/kubelet/lifecycle/handlers_test.go [hunk 1]
```diff
@@ -25,54 +25,52 @@ import (
 	"testing"
 	"time"
 
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
Comment (fromanirh) [inline pkg/kubelet/lifecycle/handlers_test.go:28]: Was this change intentional or automatically generated by the tooling?

File: pkg/kubelet/lifecycle/handlers_test.go [hunk 2]
```diff
@@ -25,54 +25,52 @@ import (
...
-			{Name: name, ContainerPort: int32(expected)},
+func TestResolvePort(t *testing.T) {
+	for _, testCase := range []struct {
+		container *v1.Container
+		name      string
+		expected  int
+	}{
+		{
+			name: "foo",
+			container: &v1.Container{
```
Comment (fromanirh) [inline pkg/kubelet/lifecycle/handlers_test.go:44]: I wonder if we should make just the `ContainerPort` object(s) a test case field (not the full `Container`)

File: pkg/kubelet/lifecycle/handlers_test.go [hunk 3]
```diff
@@ -257,3 +257,460 @@ func TestRunHandlerHttpFailure(t *testing.T) {
 		t.Errorf("unexpected url: %s", fakeHTTPGetter.url)
 	}
 }
+
+func TestNoNewPrivsRequired(t *testing.T) {
+	admitPrivEsc := false
+	for _, testCase := range []struct {
+		name     string
```
Comment (fromanirh) [inline pkg/kubelet/lifecycle/handlers_test.go:264]: this looks unused, so either please use it or drop it

File: pkg/kubelet/lifecycle/handlers_test.go [hunk 4]
```diff
@@ -257,3 +257,460 @@ func TestRunHandlerHttpFailure(t *testing.T) {
...
+		},
+	} {
+		res := noNewPrivsRequired(testCase.input)
+		if res != testCase.expected {
+			t.Errorf("expected admission result for pod to be %t, got %t", res, testCase.expected)
+		}
+	}
+}
+
+// NOTE: there _has_ to be something in kubernetes somewhere that does this already?
```
Comment (fromanirh) [inline pkg/kubelet/lifecycle/handlers_test.go:294]: `BoolPtr` in `k8s.io/utils/pointer` :)

File: pkg/kubelet/lifecycle/handlers_test.go [hunk 5]
```diff
@@ -257,3 +257,460 @@ func TestRunHandlerHttpFailure(t *testing.T) {
...
+func getBoolPointer(b bool) *bool {
+	return &b
+}
+
+func TestNoNewPrivsAdmitHandler(t *testing.T) {
```
Comment (fromanirh) [inline pkg/kubelet/lifecycle/handlers_test.go:299]: Ok, this test (and all the others) is nicely laid out, is easy to read and looks good. Could you please provide, in the commit message, as code comments o as github comment a pointer to understand the context and to learn about the expected behaviour? TL;DR: please add a reference so we can validate the tests themselves and understand the correct behaviour.

**Comment by fromanirh** [comment]
Thanks for your contribution! looks good, few inline questions mostly for better understanding

**Comment by SergeyKanzhelev** [comment]
@PI-Victor I marked the PR as "Waiting for author". Please make sure to comment once feedback is addressed.

**Comment by PI-Victor** [comment]
> @PI-Victor I marked the PR as "Waiting for author". Please make sure to comment once feedback is addressed.

Thank you for having a look, i'll address the comments asap!

**Comment by SergeyKanzhelev** [comment]
@PI-Victor any updates?

File: pkg/kubelet/lifecycle/handlers_test.go (hunk 2 above)
Comment (PI-Victor) [reply pkg/kubelet/lifecycle/handlers_test.go:44]: `resolvePort()` takes a `*v1.Container{}` and one of the test cases tests tests that port 80 should be returned if the container passed has no port specified (see line 50), it would be kinda hacky to have to account for that just by using `ContainerPort` instead.

File: pkg/kubelet/lifecycle/handlers_test.go (hunk 1 above)
Comment (PI-Victor) [reply pkg/kubelet/lifecycle/handlers_test.go]: go language server does whatever it wants, i've disabled it and removed this.

File: pkg/kubelet/lifecycle/handlers_test.go (hunk 3 above)
Comment (PI-Victor) [reply pkg/kubelet/lifecycle/handlers_test.go]: fixed

File: pkg/kubelet/lifecycle/handlers_test.go (hunk 4 above)
Comment (PI-Victor) [reply pkg/kubelet/lifecycle/handlers_test.go]: thanks a bunch! :D

File: pkg/kubelet/lifecycle/handlers_test.go (hunk 5 above)
Comment (PI-Victor) [reply pkg/kubelet/lifecycle/handlers_test.go:293]: i've addressed this in the second commit message.

**Comment by PI-Victor** [comment]
sorry for the super late reply, i've now addressed all concerns, PTAL.

**Comment by PI-Victor** [comment]
hey folks, with the freeze over, can we now merge this?

File: pkg/kubelet/lifecycle/handlers_test.go [hunk 6]
```diff
@@ -259,3 +258,453 @@ func TestRunHandlerHttpFailure(t *testing.T) {
...
+
+func TestNoNewPrivsRequired(t *testing.T) {
+	for _, testCase := range []struct {
+		input    *v1.Pod
+		expected bool
+	}{
+		{
+			input: &v1.Pod{
+				Spec: v1.PodSpec{
+					Containers: []v1.Container{
```
Comment (SergeyKanzhelev) [inline pkg/kubelet/lifecycle/handlers_test.go:270]: would be nice to test with two containers, where the first one doesn't have `AllowPrivilegeEscalation` and second does

File: pkg/kubelet/lifecycle/handlers_test.go [hunk 7]
```diff
@@ -259,3 +258,453 @@ func TestRunHandlerHttpFailure(t *testing.T) {
...
+						{
+							SecurityContext: &v1.SecurityContext{
+								AllowPrivilegeEscalation: pointer.BoolPtr(false),
```
Comment (SergeyKanzhelev) [inline pkg/kubelet/lifecycle/handlers_test.go:273]: would be nice to test when this is set to `false` explicitly

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm

**Comment by dims** [comment]
/assign @mrunalp @klueska

(for sig-node approval)

**Comment by PI-Victor** [comment]
sorry, i keep forgetting about this PR. i will address the concerns soon so it can be merged.

**Comment by endocrimes** [comment]
This would still be a good improvement without the nice-to-haves (although they would be v appreciated @PI-Victor)

/remove-lifecycle stale

**Comment by dchen1107** [comment]
/approve

There are two open comments made by @SergeyKanzhelev which are not addressed. But Sergey's comments are more on increasing the test coverage and continuing improve. Let's merge it for now.

**Comment by endocrimes** [comment]
/hold

Looks like this needs a rebase/some cleanup bc of conflicting changes

**Comment by dims** [comment]
@PI-Victor needs a rebase

**Comment by PI-Victor** [comment]
sorry, i've completely forgot about this, closing
/close
